trufflesuite/truffle
# Make dashboard bind to IPv4 & IPv6 when available

## PR description

Addresses issue #5690.

This change makes it so that the dashboard server and dashboard message bus listens on both IPv4 and IPv6 for a given hostname. In the event that one of the two protocols is not available for a given hostname, it will only listen on the protocol that is available.

## Testing instructions

1. Run `truffle dashboard` in your project of choice.
2. Ensure that you can connect to it via http://127.0.0.1:24012 (note the port number may be different on your machine)
3. Ensure that you can connect to it via http://[::1]:24012 (note the port number may be different on your machine)
4. Configure the `dashboard` network to connect to host `127.0.0.1` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.
5. Configure the `dashboard` network to connect to host `::1` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.
6. Configure the `dashboard` network to connect to host `localhost` and ensure that `truffle migrate --reset --network dashboard` succeeds under this configuration.

## Documentation


## Breaking changes and new features


## Note to reviewers

There's a bit of unfortunate code duplication here. I would have preferred to put the logic for resolving the IPs of a given local hostname into the `@truffle/dashboard-message-bus-common` package, but the webpack operation in the build for the dashboard frontend choked on the node-specific modules that are used by this logic (the `os` and `net` modules). As a result, I decided it was easier to _temporarily_ duplicate this code between the `utils` module in the dashboard server, and the `utils` module in the `dashboard-message-bus` module.

That said, we're still not breaking "the rule of three" here, so hopefully this can be approved as-is. However if #5621 happens to be ready to merge before this PR, I can rebase this code on `develop`, move the logic to `@dashboard-message-bus-common`, and drop the `net` and `os` packages in the dashboard's newly-ejected webpack config.
